bug: assert failed on oldv <= dpn_->max_i at pack_int.cpp:337

### Have you read the Contributing Guidelines on issues?

- [X] I have read the [Contributing Guidelines on issues](https://github.com/stoneatom/stonedb/blob/stonedb-5.7-dev/CONTRIBUTING.md).

### Please confirm if bug report does NOT exists already ?

- [X] I confirm there is no existing issue for this

### Describe the problem

```
[2023-04-20 13:34:30.387698] [967] [WARN] [exception.cpp:42] MSG: Exception: assert failed on oldv <= dpn_->max_i at pack_int.cpp:334, msg: [].

STACK TRACE BEGIN
        /opt/stonedb57/install//bin/mysqld(Tianmu::common::AssertException::AssertException(char const*, char const*, int, std::string const&)+0x1b7) [0x309c473]
        /opt/stonedb57/install//bin/mysqld(Tianmu::core::PackInt::DeleteByRow(unsigned long)+0x6e0) [0x3062a10]
        /opt/stonedb57/install//bin/mysqld(Tianmu::core::TianmuAttr::DeleteBatchData(Tianmu::core::Transaction*, std::vector<unsigned long, std::allocator<unsigned long> > const&)+0x411) [0x2e7bccf]
        /opt/stonedb57/install//bin/mysqld(void std::__invoke_impl<void, void (Tianmu::core::TianmuAttr::*&)(Tianmu::core::Transaction*, std::vector<unsigned long, std::allocator<unsigned long> > const&), Tianmu::core::TianmuAttr*&, Tianmu::core::Transaction*&, std::vector<unsigned long, std::allocator<unsigned long> >&>(std::__invoke_memfun_deref, void (Tianmu::core::TianmuAttr::*&)(Tianmu::core::Transaction*, std::vector<unsigned long, std::allocator<unsigned long> > const&), Tianmu::core::TianmuAttr*&, Tianmu::core::Transaction*&, std::vector<unsigned long, std::allocator<unsigned long> >&)+0x98) [0x2dc01c9]
        /opt/stonedb57/install//bin/mysqld(std::__invoke_result<void (Tianmu::core::TianmuAttr::*&)(Tianmu::core::Transaction*, std::vector<unsigned long, std::allocator<unsigned long> > const&), Tianmu::core::TianmuAttr*&, Tianmu::core::Transaction*&, std::vector<unsigned long, std::allocator<unsigned long> >&>::type std::__invoke<void (Tianmu::core::TianmuAttr::*&)(Tianmu::core::Transaction*, std::vector<unsigned long, std::allocator<unsigned long> > const&), Tianmu::core::TianmuAttr*&, Tianmu::core::Transaction*&, std::vector<unsigned long, std::allocator<unsigned long> >&>(void (Tianmu::core::TianmuAttr::*&)(Tianmu::core::Transaction*, std::vector<unsigned long, std::allocator<unsigned long> > const&), Tianmu::core::TianmuAttr*&, Tianmu::core::Transaction*&, std::vector<unsigned long, std::allocator<unsigned long> >&)+0x6f) [0x2dbfb4f]
        /opt/stonedb57/install//bin/mysqld(void std::_Bind<void (Tianmu::core::TianmuAttr::*(Tianmu::core::TianmuAttr*, Tianmu::core::Transaction*, std::vector<unsigned long, std::allocator<unsigned long> >))(Tianmu::core::Transaction*, std::vector<unsigned long, std::allocator<unsigned long> > const&)>::__call<void, , 0ul, 1ul, 2ul>(std::tuple<>&&, std::_Index_tuple<0ul, 1ul, 2ul>)+0xa0) [0x2dbf2ca]
        /opt/stonedb57/install//bin/mysqld(void std::_Bind<void (Tianmu::core::TianmuAttr::*(Tianmu::core::TianmuAttr*, Tianmu::core::Transaction*, std::vector<unsigned long, std::allocator<unsigned long> >))(Tianmu::core::Transaction*, std::vector<unsigned long, std::allocator<unsigned long> > const&)>::operator()<, void>()+0x2a) [0x2dbdb3a]
        /opt/stonedb57/install//bin/mysqld(void std::__invoke_impl<void, std::_Bind<void (Tianmu::core::TianmuAttr::*(Tianmu::core::TianmuAttr*, Tianmu::core::Transaction*, std::vector<unsigned long, std::allocator<unsigned long> >))(Tianmu::core::Transaction*, std::vector<unsigned long, std::allocator<unsigned long> > const&)>&>(std::__invoke_other, std::_Bind<void (Tianmu::core::TianmuAttr::*(Tianmu::core::TianmuAttr*, Tianmu::core::Transaction*, std::vector<unsigned long, std::allocator<unsigned long> >))(Tianmu::core::Transaction*, std::vector<unsigned long, std::allocator<unsigned long> > const&)>&)+0x20) [0x2dbc9c9]
        /opt/stonedb57/install//bin/mysqld(std::__invoke_result<std::_Bind<void (Tianmu::core::TianmuAttr::*(Tianmu::core::TianmuAttr*, Tianmu::core::Transaction*, std::vector<unsigned long, std::allocator<unsigned long> >))(Tianmu::core::Transaction*, std::vector<unsigned long, std::allocator<unsigned long> > const&)>&<> >::type std::__invoke<std::_Bind<void (Tianmu::core::TianmuAttr::*(Tianmu::core::TianmuAttr*, Tianmu::core::Transaction*, std::vector<unsigned long, std::allocator<unsigned long> >))(Tianmu::core::Transaction*, std::vector<unsigned long, std::allocator<unsigned long> > const&)>&>(std::_Bind<void (Tianmu::core::TianmuAttr::*(Tianmu::core::TianmuAttr*
[2023-04-20 13:34:30.412064] [962] [ERROR] [engine.cpp:958] MSG: Roll back query 'LOAD DELAYED INSERT DATA INTO TABLE mbs.p1jk_task'
[2023-04-20 13:34:30.412113] [962] [INFO] [transaction.cpp:90] MSG: rollback transaction 6440aed60003cdd1.
```

### Expected behavior

_No response_

### How To Reproduce

_No response_

### Environment

_No response_

### Are you interested in submitting a PR to solve the problem?

- [ ] Yes, I will!

--- a/storage/tianmu/core/delta_table.cpp
+++ b/storage/tianmu/core/delta_table.cpp
@@ -171,9 +171,9 @@ void DeltaTable::AddInsertRecord(Transaction *tx, uint64_t row_id, std::unique_p
     kv_trans.Acquiresnapshot();
     tx->ResetInsertRowNum();
   }
-  load_id++;
-  stat.write_cnt++;
-  stat.write_bytes += size;
+  load_id.fetch_add(1);
+  stat.write_cnt.fetch_add(1);
+  stat.write_bytes.fetch_add(size);
 }
 
 void DeltaTable::AddRecord(Transaction *tx, uint64_t row_id, std::unique_ptr<char[]> buf, uint32_t size) {
@@ -192,9 +192,9 @@ void DeltaTable::AddRecord(Transaction *tx, uint64_t row_id, std::unique_ptr<cha
     throw common::Exception("Error,kv_trans.PutData failed,date size: " + std::to_string(size) +
                             " date:" + std::string(buf.get()));
   }
-  load_id++;
-  stat.write_cnt++;
-  stat.write_bytes += size;
+  load_id.fetch_add(1);
+  stat.write_cnt.fetch_add(1);
+  stat.write_bytes.fetch_add(size);
 }
 
 void DeltaTable::Truncate(Transaction *tx) {

--- a/storage/tianmu/core/engine.cpp
+++ b/storage/tianmu/core/engine.cpp
@@ -291,10 +291,11 @@ int Engine::Init(uint engine_slot) {
          [this]() {
            for (auto &delta : m_table_deltas) {
              TIANMU_LOG(LogCtl_Level::INFO,
-                        "table name: %s, delta table id: %d delta_size: %ld, current load id: %ld, merge id: %ld, current row_id: %ld",
-                        delta.second->FullName().c_str(),
-                        delta.second->GetDeltaTableID(), delta.second->load_id.load() - delta.second->merge_id.load(),
-                        delta.second->load_id.load(), delta.second->merge_id.load(), delta.second->row_id.load());
+                        "table name: %s, delta table id: %d delta_size: %ld, current load id: %ld, merge id: %ld, "
+                        "current row_id: %ld",
+                        delta.second->FullName().c_str(), delta.second->GetDeltaTableID(),
+                        delta.second->load_id.load() - delta.second->merge_id.load(), delta.second->load_id.load(),
+                        delta.second->merge_id.load(), delta.second->row_id.load());
            }
          }},
         {tianmu_sysvar_log_loop_interval * 5,

--- a/storage/tianmu/core/engine.cpp
+++ b/storage/tianmu/core/engine.cpp
@@ -291,7 +291,8 @@ int Engine::Init(uint engine_slot) {
          [this]() {
            for (auto &delta : m_table_deltas) {
              TIANMU_LOG(LogCtl_Level::INFO,
-                        "delta table id: %d delta_size: %d, current load id: %d, merge id: %d, current row_id: %d",
+                        "table name: %s, delta table id: %d delta_size: %ld, current load id: %ld, merge id: %ld, current row_id: %ld",
+                        delta.second->FullName().c_str(),
                         delta.second->GetDeltaTableID(), delta.second->load_id.load() - delta.second->merge_id.load(),
                         delta.second->load_id.load(), delta.second->merge_id.load(), delta.second->row_id.load());
            }

--- a/storage/tianmu/data/pack_int.cpp
+++ b/storage/tianmu/data/pack_int.cpp
@@ -548,7 +548,8 @@ void PackInt::LoadValuesFixed(const loader::ValueCache *vc, const std::optional<
   auto delta = dpn_->min_i - new_min;
 
   if (dpn_->NullOnly() || dpn_->Uniform()) {
-    if(data_.ptr_) dealloc(data_.ptr_);
+    if (data_.ptr_)
+      dealloc(data_.ptr_);
     data_.ptr_ = alloc(new_vt * new_nr, mm::BLOCK_TYPE::BLOCK_UNCOMPRESSED);
     std::memset(data_.ptr_, 0, new_vt * new_nr);
     if (dpn_->Uniform()) {

--- a/storage/tianmu/data/pack_int.cpp
+++ b/storage/tianmu/data/pack_int.cpp
@@ -43,7 +43,7 @@ PackInt::PackInt(DPN *dpn, PackCoordinate pc, ColumnShare *s) : Pack(dpn, pc, s)
 PackInt::PackInt(const PackInt &apn, const PackCoordinate &pc) : Pack(apn, pc), is_real_(apn.is_real_) {
   data_.value_type_ = apn.data_.value_type_;
   if (apn.data_.value_type_ > 0) {
-    ASSERT(apn.data_.ptr_ != nullptr);
+    ASSERT(apn.data_.ptr_ != nullptr, "path:" + col_share_->DataFile());
     data_.ptr_ = alloc(data_.value_type_ * apn.dpn_->numOfRecords, mm::BLOCK_TYPE::BLOCK_UNCOMPRESSED);
     std::memcpy(data_.ptr_, apn.data_.ptr_, data_.value_type_ * apn.dpn_->numOfRecords);
   }
@@ -167,6 +167,7 @@ void PackInt::ExpandOrShrink(uint64_t maxv, int64_t delta) {
   auto new_vt = GetValueSize(maxv);
   if (new_vt != data_.value_type_ || delta != 0) {
     auto tmp = alloc(new_vt * dpn_->numOfRecords, mm::BLOCK_TYPE::BLOCK_UNCOMPRESSED);
+    std::memset(tmp, 0, new_vt * dpn_->numOfRecords);
     xf[{data_.value_type_, new_vt}](data_.ptr_, data_.ptr_int8_ + (data_.value_type_ * dpn_->numOfRecords), tmp, delta);
     dealloc(data_.ptr_);
     data_.ptr_ = tmp;
@@ -178,7 +179,6 @@ void PackInt::UpdateValueFloat(size_t locationInPack, const Value &v) {
   if (IsNull(locationInPack)) {
     if (!v.HasValue())
       return;
-    // ASSERT(v.HasValue(), col_share_->DataFile() + " locationInPack: " + std::to_string(locationInPack));
 
     // update null to non-null
     dpn_->synced = false;
@@ -200,7 +200,7 @@ void PackInt::UpdateValueFloat(size_t locationInPack, const Value &v) {
       SetValD(locationInPack, d);
     } else {
       dpn_->numOfNulls--;
-      ASSERT(data_.ptr_ != nullptr);
+      ASSERT(data_.ptr_ != nullptr, "path:" + col_share_->DataFile());
 
       SetValD(locationInPack, d);
 
@@ -214,7 +214,7 @@ void PackInt::UpdateValueFloat(size_t locationInPack, const Value &v) {
     // update an original non-null value
 
     if (data_.empty()) {
-      ASSERT(dpn_->Uniform());
+      ASSERT(dpn_->Uniform(), "path:" + col_share_->DataFile());
       data_.value_type_ = 8;
       data_.ptr_ = alloc(data_.value_type_ * dpn_->numOfRecords, mm::BLOCK_TYPE::BLOCK_UNCOMPRESSED);
       for (uint i = 0; i < dpn_->numOfRecords; i++) SetValD(i, dpn_->min_d);
@@ -251,7 +251,7 @@ void PackInt::UpdateValueFloat(size_t locationInPack, const Value &v) {
 
     // update non-null to another nonull
     auto newv = v.GetDouble();
-    ASSERT(oldv != newv);
+    ASSERT(oldv != newv, "path:" + col_share_->DataFile());
 
     newmin = std::min(newmin, newv);
     newmax = std::max(newmax, newv);
@@ -290,7 +290,7 @@ void PackInt::DeleteByRow(size_t locationInPack) {
   if (!IsNull(locationInPack)) {
     if (is_real_) {
       if (data_.empty()) {
-        ASSERT(dpn_->Uniform());
+        ASSERT(dpn_->Uniform(), "path:" + col_share_->DataFile());
         data_.value_type_ = 8;
         data_.ptr_ = alloc(data_.value_type_ * dpn_->numOfRecords, mm::BLOCK_TYPE::BLOCK_UNCOMPRESSED);
         for (uint i = 0; i < dpn_->numOfRecords; i++) SetValD(i, dpn_->min_d);
@@ -324,7 +324,7 @@ void PackInt::DeleteByRow(size_t locationInPack) {
       }
     } else {
       if (data_.empty()) {
-        ASSERT(dpn_->Uniform());
+        ASSERT(dpn_->Uniform(), "path:" + col_share_->DataFile());
         data_.value_type_ = GetValueSize(dpn_->max_i - dpn_->min_i);
         data_.ptr_ = alloc(data_.value_type_ * dpn_->numOfRecords, mm::BLOCK_TYPE::BLOCK_UNCOMPRESSED);
         std::memset(data_.ptr_, 0, data_.value_type_ * dpn_->numOfRecords);
@@ -376,7 +376,6 @@ void PackInt::UpdateValueFixed(size_t locationInPack, const Value &v) {
   if (IsNull(locationInPack)) {
     if (!v.HasValue())
       return;
-    // ASSERT(v.HasValue(), col_share_->DataFile() + " locationInPack: " + std::to_string(locationInPack));
     //  update null to non-null
     dpn_->synced = false;
     UnsetNull(locationInPack);
@@ -397,7 +396,7 @@ void PackInt::UpdateValueFixed(size_t locationInPack, const Value &v) {
       dpn_->numOfNulls--;
       SetVal64(locationInPack, l - dpn_->min_i);
     } else {
-      ASSERT(data_.ptr_ != nullptr);
+      ASSERT(data_.ptr_ != nullptr, "path:" + col_share_->DataFile());
       dpn_->numOfNulls--;
 
       if (l >= dpn_->min_i && l <= dpn_->max_i) {
@@ -408,31 +407,22 @@ void PackInt::UpdateValueFixed(size_t locationInPack, const Value &v) {
         return;
       }
 
-      decltype(data_.value_type_) new_vt;
       int64_t delta = 0;
       if (l < dpn_->min_i) {
         delta = dpn_->min_i - l;
         dpn_->min_i = l;
-        new_vt = GetValueSize(dpn_->max_i - dpn_->min_i);
       } else {  // so l > dpn_->max
         dpn_->max_i = l;
-        new_vt = GetValueSize(dpn_->max_i - dpn_->min_i);
       }
-      if (new_vt != data_.value_type_ || delta != 0) {
-        auto tmp = alloc(new_vt * dpn_->numOfRecords, mm::BLOCK_TYPE::BLOCK_UNCOMPRESSED);
-        xf[{data_.value_type_, new_vt}](data_.ptr_, data_.ptr_int8_ + (data_.value_type_ * dpn_->numOfRecords), tmp,
-                                        delta);
-        dealloc(data_.ptr_);
-        data_.ptr_ = tmp;
-        data_.value_type_ = new_vt;
-      }
+
+      ExpandOrShrink(dpn_->max_i - dpn_->min_i, delta);
       SetVal64(locationInPack, l - dpn_->min_i);
     }
   } else {
     // update an original non-null value
 
     if (data_.empty()) {
-      ASSERT(dpn_->Uniform());
+      ASSERT(dpn_->Uniform(), "path:" + col_share_->DataFile());
       data_.value_type_ = GetValueSize(dpn_->max_i - dpn_->min_i);
       data_.ptr_ = alloc(data_.value_type_ * dpn_->numOfRecords, mm::BLOCK_TYPE::BLOCK_UNCOMPRESSED);
       std::memset(data_.ptr_, 0, data_.value_type_ * dpn_->numOfRecords);
@@ -501,17 +491,11 @@ void PackInt::LoadValuesDouble(const loader::ValueCache *vc, const std::optional
   auto new_min = std::min(vc->MinDouble(), dpn_->min_d);
   auto new_max = std::max(vc->MaxDouble(), dpn_->max_d);
   auto new_nr = dpn_->numOfRecords + vc->NumOfValues();
-  if (dpn_->NullOnly() || dpn_->Uniform()) {
-    ASSERT(data_.ptr_ == nullptr);
-    data_.value_type_ = sizeof(double);
-    data_.ptr_ = alloc(data_.value_type_ * new_nr, mm::BLOCK_TYPE::BLOCK_UNCOMPRESSED);
-    if (dpn_->Uniform()) {
-      std::fill_n(reinterpret_cast<double *>(data_.ptr_), new_nr, dpn_->min_d);
-    }
-  } else {
-    // expanding existing pack data_
-    ASSERT(data_.ptr_ != nullptr);
-    data_.ptr_ = rc_realloc(data_.ptr_, data_.value_type_ * new_nr, mm::BLOCK_TYPE::BLOCK_UNCOMPRESSED);
+  data_.value_type_ = sizeof(double);
+
+  data_.ptr_ = rc_realloc(data_.ptr_, data_.value_type_ * new_nr, mm::BLOCK_TYPE::BLOCK_UNCOMPRESSED);
+  if (dpn_->Uniform()) {
+    std::fill_n(reinterpret_cast<double *>(data_.ptr_), new_nr, dpn_->min_d);
   }
 
   dpn_->synced = false;
@@ -558,13 +542,15 @@ void PackInt::LoadValuesFixed(const loader::ValueCache *vc, const std::optional<
   auto new_vt = GetValueSize(new_max - new_min);
   auto new_nr = dpn_->numOfRecords + vc->NumOfValues();
 
-  ASSERT(new_vt >= data_.value_type_);
+  ASSERT(new_vt >= data_.value_type_, col_share_->DataFile() + ", new_vt=:" + std::to_string(new_vt) +
+                                          " , data_.value_type_:" + std::to_string(data_.value_type_));
 
   auto delta = dpn_->min_i - new_min;
 
   if (dpn_->NullOnly() || dpn_->Uniform()) {
-    ASSERT(data_.ptr_ == nullptr);
+    if(data_.ptr_) dealloc(data_.ptr_);
     data_.ptr_ = alloc(new_vt * new_nr, mm::BLOCK_TYPE::BLOCK_UNCOMPRESSED);
+    std::memset(data_.ptr_, 0, new_vt * new_nr);
     if (dpn_->Uniform()) {
       switch (new_vt) {
         case 8:
@@ -586,8 +572,10 @@ void PackInt::LoadValuesFixed(const loader::ValueCache *vc, const std::optional<
     }
   } else {
     // expanding existing pack data_
-    ASSERT(data_.ptr_ != nullptr);
+    ASSERT(data_.ptr_ != nullptr, "path:" + col_share_->DataFile());
+
     auto tmp_ptr = alloc(new_vt * new_nr, mm::BLOCK_TYPE::BLOCK_UNCOMPRESSED);
+    std::memset(tmp_ptr, 0, new_vt * new_nr);
     xf[{data_.value_type_, new_vt}](data_.ptr_, data_.ptr_int8_ + (data_.value_type_ * dpn_->numOfRecords), tmp_ptr,
                                     delta);
     dealloc(data_.ptr_);
@@ -602,10 +590,11 @@ void PackInt::LoadValuesFixed(const loader::ValueCache *vc, const std::optional<
     if (vc->NotNull(i)) {
       AppendValue(*(reinterpret_cast<uint64_t *>(const_cast<char *>(vc->GetDataBytesPointer(i)))) - new_min);
     } else {
-      if (nv.has_value())
+      if (nv.has_value()) {
         AppendValue(nv->i - new_min);
-      else
+      } else {
         AppendNull();
+      }
       if (vc->IsDelete(i)) {
         if (!IsNull(dpn_->numOfRecords - 1)) {
           SetNull(dpn_->numOfRecords - 1);
@@ -616,6 +605,7 @@ void PackInt::LoadValuesFixed(const loader::ValueCache *vc, const std::optional<
       }
     }
   }
+
   // sum has already been updated outside
   dpn_->min_i = new_min;
   dpn_->max_i = new_max;
@@ -656,7 +646,7 @@ void PackInt::LoadDataFromFile(system::Stream *fcurfile) {
     fcurfile->ReadExact(uptr.get(), dpn_->dataLength);
     dpn_->synced = true;
     {
-      ASSERT(!IsModeNoCompression());
+      ASSERT(!IsModeNoCompression(), "path:" + col_share_->DataFile());
 
       uint *cur_buf = static_cast<uint *>(uptr.get());
       if (data_.ptr_ == nullptr && data_.value_type_ > 0)
@@ -758,7 +748,7 @@ void PackInt::Save() {
   else
     f.WriteExact(uptr.get(), dpn_->dataLength);
 
-  ASSERT(f.Tell() == off_t(dpn_->dataAddress + dpn_->dataLength));
+  ASSERT(f.Tell() == off_t(dpn_->dataAddress + dpn_->dataLength), "path:" + col_share_->DataFile());
 
   f.Close();
   dpn_->synced = true;

--- a/storage/tianmu/data/pack_int.h
+++ b/storage/tianmu/data/pack_int.h
@@ -72,19 +72,21 @@ class PackInt final : public Pack {
   }
 
   void AppendNull() {
-    SetNull(dpn_->numOfRecords);
-    dpn_->numOfNulls++;
     dpn_->numOfRecords++;
+    if(likely(NotNull(dpn_->numOfRecords-1))){
+      SetNull(dpn_->numOfRecords-1);
+    }
+    dpn_->numOfNulls++;
   }
   void SetValD(uint n, double v) {
     dpn_->synced = false;
-    ASSERT(n < dpn_->numOfRecords);
+    ASSERT(n < dpn_->numOfRecords , "n: "+ std::to_string(n) + " ,dpn->numOfRecords: "+ std::to_string(dpn_->numOfRecords));
     ASSERT(is_real_);
     data_.ptr_double_[n] = v;
   }
   void SetVal64(uint n, uint64_t v) {
     dpn_->synced = false;
-    ASSERT(n < dpn_->numOfRecords);
+    ASSERT(n < dpn_->numOfRecords , "n: "+ std::to_string(n) + " ,dpn->numOfRecords: "+ std::to_string(dpn_->numOfRecords));
     switch (data_.value_type_) {
       case 8:
         data_.ptr_int64_[n] = v;

--- a/storage/tianmu/data/pack_int.h
+++ b/storage/tianmu/data/pack_int.h
@@ -73,20 +73,22 @@ class PackInt final : public Pack {
 
   void AppendNull() {
     dpn_->numOfRecords++;
-    if(likely(NotNull(dpn_->numOfRecords-1))){
-      SetNull(dpn_->numOfRecords-1);
+    if (likely(NotNull(dpn_->numOfRecords - 1))) {
+      SetNull(dpn_->numOfRecords - 1);
     }
     dpn_->numOfNulls++;
   }
   void SetValD(uint n, double v) {
     dpn_->synced = false;
-    ASSERT(n < dpn_->numOfRecords , "n: "+ std::to_string(n) + " ,dpn->numOfRecords: "+ std::to_string(dpn_->numOfRecords));
+    ASSERT(n < dpn_->numOfRecords,
+           "n: " + std::to_string(n) + " ,dpn->numOfRecords: " + std::to_string(dpn_->numOfRecords));
     ASSERT(is_real_);
     data_.ptr_double_[n] = v;
   }
   void SetVal64(uint n, uint64_t v) {
     dpn_->synced = false;
-    ASSERT(n < dpn_->numOfRecords , "n: "+ std::to_string(n) + " ,dpn->numOfRecords: "+ std::to_string(dpn_->numOfRecords));
+    ASSERT(n < dpn_->numOfRecords,
+           "n: " + std::to_string(n) + " ,dpn->numOfRecords: " + std::to_string(dpn_->numOfRecords));
     switch (data_.value_type_) {
       case 8:
         data_.ptr_int64_[n] = v;

--- a/storage/tianmu/data/pack_str.cpp
+++ b/storage/tianmu/data/pack_str.cpp
@@ -226,13 +226,14 @@ void PackStr::UpdateValue(size_t locationInPack, const Value &v) {
       SetPtrSize(locationInPack, nullptr, 0);
       SetNull(locationInPack);
       dpn_->numOfNulls++;
-      if(dpn_->NullOnly()){
+      if (dpn_->NullOnly()) {
         data_.sum_len = 0;
       }
     } else {
       // update non-null to another nonull
       auto vsize = GetValueBinary(locationInPack).size();
-      ASSERT(data_.sum_len >= vsize, col_share_->DataFile() + ", data_.sum_len:" + std::to_string(data_.sum_len) + ", vsize:"+std::to_string(vsize));
+      ASSERT(data_.sum_len >= vsize, col_share_->DataFile() + ", data_.sum_len:" + std::to_string(data_.sum_len) +
+                                         ", vsize:" + std::to_string(vsize));
       data_.sum_len -= vsize;
       auto &str = v.GetString();
       if (str.size() <= vsize) {
@@ -262,7 +263,7 @@ void PackStr::DeleteByRow(size_t locationInPack) {
     SetPtrSize(locationInPack, nullptr, 0);
     SetNull(locationInPack);
     dpn_->numOfNulls++;
-    if(dpn_->NullOnly()){
+    if (dpn_->NullOnly()) {
       data_.sum_len = 0;
     }
   }
@@ -386,8 +387,8 @@ std::pair<PackStr::UniquePtr, size_t> PackStr::Compress() {
         reinterpret_cast<uint *>(alloc(comp_len_buf_size / 4 * sizeof(uint), mm::BLOCK_TYPE::BLOCK_TEMPORARY)), *this);
     uint tmp_comp_len_buf_size = comp_len_buf_size - 8;
     compress::NumCompressor<uint> nc;
-    CprsErr res = nc.Compress(reinterpret_cast<char *>(comp_len_buf.get() + 2), tmp_comp_len_buf_size,
-                              nc_buffer.get(), onn, maxv);
+    CprsErr res = nc.Compress(reinterpret_cast<char *>(comp_len_buf.get() + 2), tmp_comp_len_buf_size, nc_buffer.get(),
+                              onn, maxv);
     if (res != CprsErr::CPRS_SUCCESS) {
       throw common::InternalException("Compression of lengths of values failed for column " +
                                       std::to_string(pc_column(GetCoordinate().co.pack) + 1) + ", pack " +
@@ -564,8 +565,9 @@ void PackStr::Save() {
     SaveUncompressed(&f);
   }
 
-  ASSERT(f.Tell() == off_t(dpn_->dataAddress + dpn_->dataLength), col_share_->DataFile() + ", " +
-         std::to_string(dpn_->dataAddress) + ":" + std::to_string(dpn_->dataLength) + "/" + std::to_string(f.Tell()));
+  ASSERT(f.Tell() == off_t(dpn_->dataAddress + dpn_->dataLength),
+         col_share_->DataFile() + ", " + std::to_string(dpn_->dataAddress) + ":" + std::to_string(dpn_->dataLength) +
+             "/" + std::to_string(f.Tell()));
   dpn_->synced = true;
 }
 
@@ -573,10 +575,9 @@ void PackStr::SaveUncompressed(system::Stream *f) {
   f->WriteExact(nulls_ptr_.get(), bitmap_size_);
   f->WriteExact(deletes_ptr_.get(), bitmap_size_);
   f->WriteExact(data_.lens, (data_.len_mode * (1 << col_share_->pss)));
-  if (data_.v.empty() || dpn_->NullOnly()){
+  if (data_.v.empty() || dpn_->NullOnly()) {
     return;
   }
-
 
   std::unique_ptr<char[]> buff(new char[data_.sum_len]);
   char *ptr = buff.get();
@@ -593,7 +594,7 @@ void PackStr::SaveUncompressed(system::Stream *f) {
 }
 
 void PackStr::LoadCompressed(system::Stream *f) {
-  ASSERT(IsModeCompressionApplied() , "path:" + col_share_->DataFile());
+  ASSERT(IsModeCompressionApplied(), "path:" + col_share_->DataFile());
 
   auto compressed_buf = alloc_ptr(dpn_->dataLength + 1, mm::BLOCK_TYPE::BLOCK_COMPRESSED);
   f->ReadExact(compressed_buf.get(), dpn_->dataLength);
@@ -634,8 +635,9 @@ void PackStr::LoadCompressed(system::Stream *f) {
       std::memcpy(deletes_ptr_.get(), reinterpret_cast<char *>(cur_buf) + sizeof(ushort), delete_buf_size);
     else {
       compress::BitstreamCompressor bsc;
-      CprsErr res = bsc.Decompress(reinterpret_cast<char *>(deletes_ptr_.get()), delete_buf_size,
-                                   reinterpret_cast<char *>(cur_buf) + sizeof(ushort), dpn_->numOfRecords, dpn_->numOfDeleted);
+      CprsErr res =
+          bsc.Decompress(reinterpret_cast<char *>(deletes_ptr_.get()), delete_buf_size,
+                         reinterpret_cast<char *>(cur_buf) + sizeof(ushort), dpn_->numOfRecords, dpn_->numOfDeleted);
       if (res != CprsErr::CPRS_SUCCESS) {
         throw common::DatabaseException("Decompression of deletes failed for column " +
                                         std::to_string(pc_column(GetCoordinate().co.pack) + 1) + ", pack " +
@@ -653,8 +655,8 @@ void PackStr::LoadCompressed(system::Stream *f) {
     compress::NumCompressor<uint> nc;
     mm::MMGuard<uint> cn_ptr((uint *)alloc((1 << col_share_->pss) * sizeof(uint), mm::BLOCK_TYPE::BLOCK_TEMPORARY),
                              *this);
-    CprsErr res = nc.Decompress(cn_ptr.get(), reinterpret_cast<char *>(cur_buf + sizeof(uint32_t)*2), comp_len_buf_size - 8,
-                                dpn_->numOfRecords - dpn_->numOfNulls, maxv);
+    CprsErr res = nc.Decompress(cn_ptr.get(), reinterpret_cast<char *>(cur_buf + sizeof(uint32_t) * 2),
+                                comp_len_buf_size - sizeof(uint32_t) * 2, dpn_->numOfRecords - dpn_->numOfNulls, maxv);
     if (res != CprsErr::CPRS_SUCCESS) {
       std::stringstream msg_buf;
       msg_buf << "Decompression of lengths of std::string values failed for column "
@@ -675,7 +677,7 @@ void PackStr::LoadCompressed(system::Stream *f) {
   cur_buf += comp_len_buf_size;
 
   auto dlen = *reinterpret_cast<uint32_t *>(cur_buf);
-  cur_buf += sizeof(dlen);
+  cur_buf += sizeof(uint32_t);
   data_.sum_len = *reinterpret_cast<uint32_t *>(cur_buf);
   cur_buf += sizeof(uint32_t);
 
@@ -721,7 +723,7 @@ void PackStr::LoadCompressed(system::Stream *f) {
 }
 
 void PackStr::LoadCompressedTrie(system::Stream *f) {
-  ASSERT(IsModeCompressionApplied() , "path:" + col_share_->DataFile());
+  ASSERT(IsModeCompressionApplied(), "path:" + col_share_->DataFile());
   compressed_data_.reset(nullptr);
   compressed_data_ = alloc_ptr(dpn_->dataLength + 1, mm::BLOCK_TYPE::BLOCK_COMPRESSED);
   f->ReadExact(compressed_data_.get(), dpn_->dataLength);

--- a/storage/tianmu/data/pack_str.cpp
+++ b/storage/tianmu/data/pack_str.cpp
@@ -205,7 +205,6 @@ void PackStr::UpdateValue(size_t locationInPack, const Value &v) {
       dpn_->max_i = -1;
     }
 
-    // ASSERT(v.HasValue(), col_share_->DataFile() + " locationInPack: " + std::to_string(locationInPack));
     UnsetNull(locationInPack);
     dpn_->numOfNulls--;
     auto &str = v.GetString();
@@ -227,10 +226,13 @@ void PackStr::UpdateValue(size_t locationInPack, const Value &v) {
       SetPtrSize(locationInPack, nullptr, 0);
       SetNull(locationInPack);
       dpn_->numOfNulls++;
+      if(dpn_->NullOnly()){
+        data_.sum_len = 0;
+      }
     } else {
       // update non-null to another nonull
       auto vsize = GetValueBinary(locationInPack).size();
-      ASSERT(data_.sum_len >= vsize);
+      ASSERT(data_.sum_len >= vsize, col_share_->DataFile() + ", data_.sum_len:" + std::to_string(data_.sum_len) + ", vsize:"+std::to_string(vsize));
       data_.sum_len -= vsize;
       auto &str = v.GetString();
       if (str.size() <= vsize) {
@@ -260,6 +262,9 @@ void PackStr::DeleteByRow(size_t locationInPack) {
     SetPtrSize(locationInPack, nullptr, 0);
     SetNull(locationInPack);
     dpn_->numOfNulls++;
+    if(dpn_->NullOnly()){
+      data_.sum_len = 0;
+    }
   }
   SetDeleted(locationInPack);
   dpn_->numOfDeleted++;
@@ -381,7 +386,7 @@ std::pair<PackStr::UniquePtr, size_t> PackStr::Compress() {
         reinterpret_cast<uint *>(alloc(comp_len_buf_size / 4 * sizeof(uint), mm::BLOCK_TYPE::BLOCK_TEMPORARY)), *this);
     uint tmp_comp_len_buf_size = comp_len_buf_size - 8;
     compress::NumCompressor<uint> nc;
-    CprsErr res = nc.Compress(reinterpret_cast<char *>(comp_len_buf.get() + sizeof(ushort)), tmp_comp_len_buf_size,
+    CprsErr res = nc.Compress(reinterpret_cast<char *>(comp_len_buf.get() + 2), tmp_comp_len_buf_size,
                               nc_buffer.get(), onn, maxv);
     if (res != CprsErr::CPRS_SUCCESS) {
       throw common::InternalException("Compression of lengths of values failed for column " +
@@ -409,7 +414,7 @@ std::pair<PackStr::UniquePtr, size_t> PackStr::Compress() {
 
   mm::MMGuard<char> comp_buf(reinterpret_cast<char *>(alloc(dlen, mm::BLOCK_TYPE::BLOCK_TEMPORARY)), *this);
 
-  if (data_.sum_len) {
+  if (data_.sum_len > 0) {
     int objs = (dpn_->numOfRecords - dpn_->numOfNulls) - zlo;
 
     mm::MMGuard<char *> tmp_index(
@@ -559,7 +564,7 @@ void PackStr::Save() {
     SaveUncompressed(&f);
   }
 
-  ASSERT(f.Tell() == off_t(dpn_->dataAddress + dpn_->dataLength),
+  ASSERT(f.Tell() == off_t(dpn_->dataAddress + dpn_->dataLength), col_share_->DataFile() + ", " +
          std::to_string(dpn_->dataAddress) + ":" + std::to_string(dpn_->dataLength) + "/" + std::to_string(f.Tell()));
   dpn_->synced = true;
 }
@@ -568,8 +573,10 @@ void PackStr::SaveUncompressed(system::Stream *f) {
   f->WriteExact(nulls_ptr_.get(), bitmap_size_);
   f->WriteExact(deletes_ptr_.get(), bitmap_size_);
   f->WriteExact(data_.lens, (data_.len_mode * (1 << col_share_->pss)));
-  if (data_.v.empty())
+  if (data_.v.empty() || dpn_->NullOnly()){
     return;
+  }
+
 
   std::unique_ptr<char[]> buff(new char[data_.sum_len]);
   char *ptr = buff.get();
@@ -586,7 +593,7 @@ void PackStr::SaveUncompressed(system::Stream *f) {
 }
 
 void PackStr::LoadCompressed(system::Stream *f) {
-  ASSERT(IsModeCompressionApplied());
+  ASSERT(IsModeCompressionApplied() , "path:" + col_share_->DataFile());
 
   auto compressed_buf = alloc_ptr(dpn_->dataLength + 1, mm::BLOCK_TYPE::BLOCK_COMPRESSED);
   f->ReadExact(compressed_buf.get(), dpn_->dataLength);
@@ -619,7 +626,7 @@ void PackStr::LoadCompressed(system::Stream *f) {
   }
 
   if (dpn_->numOfDeleted > 0) {
-    uint delete_buf_size = 0;
+    ushort delete_buf_size = 0;
     delete_buf_size = (*reinterpret_cast<ushort *>(cur_buf));
     if (delete_buf_size > bitmap_size_)
       throw common::DatabaseException("Unexpected bytes found in data pack.");
@@ -628,7 +635,7 @@ void PackStr::LoadCompressed(system::Stream *f) {
     else {
       compress::BitstreamCompressor bsc;
       CprsErr res = bsc.Decompress(reinterpret_cast<char *>(deletes_ptr_.get()), delete_buf_size,
-                                   reinterpret_cast<char *>(cur_buf) + 2, dpn_->numOfRecords, dpn_->numOfDeleted);
+                                   reinterpret_cast<char *>(cur_buf) + sizeof(ushort), dpn_->numOfRecords, dpn_->numOfDeleted);
       if (res != CprsErr::CPRS_SUCCESS) {
         throw common::DatabaseException("Decompression of deletes failed for column " +
                                         std::to_string(pc_column(GetCoordinate().co.pack) + 1) + ", pack " +
@@ -646,7 +653,7 @@ void PackStr::LoadCompressed(system::Stream *f) {
     compress::NumCompressor<uint> nc;
     mm::MMGuard<uint> cn_ptr((uint *)alloc((1 << col_share_->pss) * sizeof(uint), mm::BLOCK_TYPE::BLOCK_TEMPORARY),
                              *this);
-    CprsErr res = nc.Decompress(cn_ptr.get(), reinterpret_cast<char *>(cur_buf + 8), comp_len_buf_size - 8,
+    CprsErr res = nc.Decompress(cn_ptr.get(), reinterpret_cast<char *>(cur_buf + sizeof(uint32_t)*2), comp_len_buf_size - 8,
                                 dpn_->numOfRecords - dpn_->numOfNulls, maxv);
     if (res != CprsErr::CPRS_SUCCESS) {
       std::stringstream msg_buf;
@@ -714,7 +721,7 @@ void PackStr::LoadCompressed(system::Stream *f) {
 }
 
 void PackStr::LoadCompressedTrie(system::Stream *f) {
-  ASSERT(IsModeCompressionApplied());
+  ASSERT(IsModeCompressionApplied() , "path:" + col_share_->DataFile());
   compressed_data_.reset(nullptr);
   compressed_data_ = alloc_ptr(dpn_->dataLength + 1, mm::BLOCK_TYPE::BLOCK_COMPRESSED);
   f->ReadExact(compressed_data_.get(), dpn_->dataLength);

--- a/storage/tianmu/executor/merge_operator.cpp
+++ b/storage/tianmu/executor/merge_operator.cpp
@@ -169,16 +169,7 @@ bool RecordMergeOperator::Merge(const rocksdb::Slice &key, const rocksdb::Slice 
       return false;
     }
   } else if (existing_type == RecordType::kDelete) {
-    DeltaRecordHeadForDelete e_deleteRecord;
-    e_deleteRecord.recordDecode(e_ptr);
-
-    DeltaRecordHeadForDelete deleteRecord;
-    deleteRecord.recordDecode(ptr);
-
-    uint32_t n_load_num = e_deleteRecord.load_num_ + deleteRecord.load_num_;
-    DeltaRecordHeadForDelete n_deleteRecord(n_load_num);
-    n_ptr = n_deleteRecord.recordEncode(n_ptr);
-    *new_value = std::string(value_buff.get(), n_ptr - value_buff.get());
+    *new_value = existing_value->ToString();
     return true;
   } else {
     return false;

--- a/storage/tianmu/handler/ha_tianmu.cpp
+++ b/storage/tianmu/handler/ha_tianmu.cpp
@@ -2016,8 +2016,6 @@ char *strmov_str(char *dst, const char *src) {
 static int tianmu_done_func([[maybe_unused]] void *p) {
   DBUG_ENTER(__PRETTY_FUNCTION__);
 
-  assert(p == tianmu_hton);
-
   if (tianmu_hton->data) {
     core::Engine *eng = reinterpret_cast<core::Engine *>(tianmu_hton->data);
     delete eng;

